src/adapter/src/catalog/storage.rs

```diff
@@ -690,46 +636,116 @@ impl<S: Append> Connection<S> {
     }
 
     pub async fn transaction<'a>(&'a mut self) -> Result<Transaction<'a, S>, Error> {
-        let databases = COLLECTION_DATABASE.peek_one(&mut self.stash).await?;
-        let schemas = COLLECTION_SCHEMA.peek_one(&mut self.stash).await?;
-        let roles = COLLECTION_ROLE.peek_one(&mut self.stash).await?;
-        let items = COLLECTION_ITEM.peek_one(&mut self.stash).await?;
-        let compute_instances = COLLECTION_COMPUTE_INSTANCES
-            .peek_one(&mut self.stash)
-            .await?;
-        let compute_instance_replicas = COLLECTION_COMPUTE_INSTANCE_REPLICAS
-            .peek_one(&mut self.stash)
-            .await?;
-        let introspection_sources = COLLECTION_COMPUTE_INTROSPECTION_SOURCE_INDEX
-            .peek_one(&mut self.stash)
-            .await?;
-        let id_allocator = COLLECTION_ID_ALLOC.peek_one(&mut self.stash).await?;
-
-        Ok(Transaction {
-            stash: &mut self.stash,
-            databases: TableTransaction::new(databases, |a, b| a.name == b.name),
-            schemas: TableTransaction::new(schemas, |a, b| {
-                a.database_id == b.database_id && a.name == b.name
-            }),
-            items: TableTransaction::new(items, |a, b| {
-                a.schema_id == b.schema_id && a.name == b.name
-            }),
-            roles: TableTransaction::new(roles, |a, b| a.name == b.name),
-            compute_instances: TableTransaction::new(compute_instances, |a, b| a.name == b.name),
-            compute_instance_replicas: TableTransaction::new(compute_instance_replicas, |a, b| {
-                a.compute_instance_id == b.compute_instance_id && a.name == b.name
-            }),
-            introspection_sources: TableTransaction::new(introspection_sources, |_a, _b| false),
-            id_allocator: TableTransaction::new(id_allocator, |_a, _b| false),
-            audit_log_updates: Vec::new(),
-        })
+        transaction(&mut self.stash).await
     }
 
     pub fn cluster_id(&self) -> Uuid {
         self.cluster_id
     }
 }
 
+/// Same as [`transaction`], except it will read an empty collection if it fails to read an initial
+/// collection state. This is because a collection is not readable until it is written to for the
+/// first time.
+///
+/// This method should only be used during migrations, when Stash may have never been written to.
+pub async fn transaction_empty<'a, S: Append>(
```

Comment:
I'm not user happy about this approach, but I'm not sure if there's a better one. Maybe collection uppers should just be initialized to one higher than the collection since. That way every collection is immediately readable with an empty value.

Reply:
I think that should work. Instead of tracking what version each collection came from, we could just go through all of them on startup and initialize any of them that have an upper of timestamp min.